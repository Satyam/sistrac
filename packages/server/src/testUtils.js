import { format } from 'mysql';
import { Router as createRouter } from 'express';

import { init as initDb } from './dbOps';
import initRouter from './restServers';

export class MysqlError extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
  }
}
export class MockDb {
  constructor(result) {
    this.result = result;
  }
  query(...args) {
    return new Promise((resolve, reject) => {
      this.sql = format(...args);
      if (this.result instanceof MysqlError) {
        reject(this.result);
      } else resolve(this.result);
    });
  }
}
export const testMethod = (
  method,
  queryResult,
  methodResult,
  ...args
) => async done => {
  const mockDb = new MockDb(queryResult);
  initDb(mockDb);
  const query = await method(...args);
  expect(query).toEqual(methodResult);
  expect(mockDb.sql).toMatchSnapshot();
  done();
};

export class Response {
  constructor(done) {
    this.done = done;
    this.statusCode = 200;
    this.cookies = {};
  }
  status(value) {
    this.statusCode = value;
    return this;
  }
  json(value) {
    this.done(this.statusCode, value);
  }
  send(value) {
    this.done(this.statusCode, value);
  }
  end() {
    this.done(this.statusCode);
  }
  cookie(name, value, opts) {
    this.cookies[name] = value;
  }
  clearCookie(name) {
    delete this.cookies[name];
  }
}

let dataRouter;
export function initRoutes() {
  dataRouter = createRouter();
  initRouter(dataRouter);
}

export const testREST = (
  method,
  path,
  queryResult,
  status,
  restResult,
  body,
) => async done => {
  const mockDb = new MockDb(queryResult);
  const res = new Response((st, resp) => {
    expect(st).toBe(status);
    expect(resp).toEqual(restResult);
    expect(mockDb.sql).toMatchSnapshot();
    done();
  });
  initDb(mockDb);
  dataRouter(
    {
      method,
      url: path,
      baseUrl: '/',
      path,
      body,
    },
    res,
    () => {
      expect('Next should not be called').toBeFalsy();
    },
  );
};

export const GET = 'GET';
export const PUT = 'PUT';
export const POST = 'POST';
export const DELETE = 'DELETE';
