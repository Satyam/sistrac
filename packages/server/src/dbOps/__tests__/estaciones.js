import { init, close } from '../';

import { MysqlError, testMethod } from '../../testUtils';

import {
  readEstaciones,
  createEstacion,
  existeEstacionPorNombre,
  existeEstacionPorSigla,
  updateEstacion,
  readEstacion,
  readEstacionPorNombre,
  readEstacionPorSigla,
  deleteEstacion,
} from '../estaciones';

const NOMBRE_ESTACION = 'YYYY';
const SIGLA_ESTACION = 'YYY';

describe('dbOps.estaciones', () => {
  describe('with mock db', () => {
    const retiro = {
      nombre: 'Retiro',
      sigla: 'RET',
    };
    it(
      'readEstaciones',
      testMethod(readEstaciones, [{ esto: 'aquello' }], [{ esto: 'aquello' }]),
    );
    it(
      'createEstacion success',
      testMethod(createEstacion, { insertId: 42 }, 42, retiro),
    );
    it(
      'createEstacion fail duplicate record',
      testMethod(createEstacion, new MysqlError('ER_DUP_ENTRY'), false, retiro),
    );
    // it('createEstacion fail other error', () => {
    //   debugger;
    //   testMethod(createEstacion, new MysqlError('OTHER_ERROR'), false, retiro);
    // });
    it(
      'existeEstacionPorNombre success',
      testMethod(existeEstacionPorNombre, [{ hay: 1 }], true, 'Retiro'),
    );
    it(
      'existeEstacionPorNombre fail',
      testMethod(existeEstacionPorNombre, [{ hay: 0 }], false, 'Retiro'),
    );
    it(
      'existeEstacionPorSigla success',
      testMethod(existeEstacionPorSigla, [{ hay: 1 }], true, 'RET'),
    );
    it(
      'existeEstacionPorSigla fail',
      testMethod(existeEstacionPorSigla, [{ hay: 0 }], false, 'RET'),
    );
    it(
      'updateEstacion success ',
      testMethod(updateEstacion, { affectedRows: 1 }, true, 25, {
        nombre: 'Retiro',
        sigla: 'RET',
      }),
    );
    it(
      'updateEstacion fail ',
      testMethod(updateEstacion, { affectedRows: 0 }, false, 25, retiro),
    );
    it(
      'readEstacion',
      testMethod(readEstacion, [retiro], retiro, { idEstacion: 25 }),
    );
    it(
      'readEstacionPorNombre',
      testMethod(readEstacionPorNombre, [retiro], retiro, retiro.nombre),
    );
    it(
      'readEstacionPorSigla',
      testMethod(readEstacionPorSigla, [retiro], retiro, retiro.sigla),
    );
    it(
      'deleteEstacion success',
      testMethod(deleteEstacion, { affectedRows: 1 }, true, 25),
    );
    it(
      'deleteEstacion fail',
      testMethod(deleteEstacion, { affectedRows: 0 }, false, 25),
    );
  });
  describe('with actual DB', () => {
    beforeAll(() => init());
    afterAll(() => close());
    it('They should all be functions', () => {
      expect(typeof readEstaciones).toBe('function');
      expect(typeof createEstacion).toBe('function');
      expect(typeof existeEstacionPorNombre).toBe('function');
      expect(typeof existeEstacionPorSigla).toBe('function');
      expect(typeof updateEstacion).toBe('function');
      expect(typeof readEstacion).toBe('function');
      expect(typeof readEstacionPorNombre).toBe('function');
      expect(typeof readEstacionPorSigla).toBe('function');
      expect(typeof deleteEstacion).toBe('function');
    });
    describe('readEstaciones', () => {
      it('should return a list', async () => {
        const query = readEstaciones();
        expect(query).toBeInstanceOf(Promise);
        const rows = await query;
        expect(Array.isArray(rows)).toBeTruthy();
        rows.forEach(row => {
          expect(row).toEqual(
            expect.objectContaining({
              idEstacion: expect.any(Number),
              nombre: expect.any(String),
              latitud: expect.any(Number),
              longitud: expect.any(Number),
              sigla: expect.any(String),
            }),
          );
          expect(row.sigla.length).toBe(3);
          expect(row.nombre.length).toBeGreaterThan(0);
        });
      });
    });
    describe('Con nueva estacion', () => {
      let idEstacion;
      beforeAll(async () => {
        const query = createEstacion({
          nombre: NOMBRE_ESTACION,
          sigla: SIGLA_ESTACION,
          latitud: 25,
          longitud: 52,
        });
        expect(query).toBeInstanceOf(Promise);
        const result = await query;
        expect(typeof result).toBe('number');
        idEstacion = result;
      });
      afterAll(async () => {
        const query = deleteEstacion(idEstacion);
        expect(query).toBeInstanceOf(Promise);
        const result = await query;
        expect(result).toBeTruthy();
      });
      describe('readEstacion', () => {
        it('by id:should return the one asked for', async () => {
          const query = readEstacion(idEstacion);
          expect(query).toBeInstanceOf(Promise);
          const row = await query;
          expect(row.idEstacion).toBe(idEstacion);
          expect(row.nombre).toBe(NOMBRE_ESTACION);
          expect(row.sigla).toBe(SIGLA_ESTACION);
        });
        it('by name: should return the one asked for', async () => {
          const query = readEstacionPorNombre(NOMBRE_ESTACION);
          expect(query).toBeInstanceOf(Promise);
          const row = await query;
          expect(row.idEstacion).toBe(idEstacion);
          expect(row.nombre).toBe(NOMBRE_ESTACION);
          expect(row.sigla).toBe(SIGLA_ESTACION);
        });
        it('by sigla: should return the one asked for', async () => {
          const query = readEstacionPorSigla(SIGLA_ESTACION);
          expect(query).toBeInstanceOf(Promise);
          const row = await query;
          expect(row.idEstacion).toBe(idEstacion);
          expect(row.nombre).toBe(NOMBRE_ESTACION);
          expect(row.sigla).toBe(SIGLA_ESTACION);
        });
        it('should fail on non-existing', async () => {
          const query = readEstacion(0);
          expect(query).toBeInstanceOf(Promise);
          const row = await query;
          expect(row).toBeUndefined();
        });
      });
      describe('existeEstacion', () => {
        it('YYYY should exist', async () => {
          const query = existeEstacionPorNombre(NOMBRE_ESTACION);
          expect(query).toBeInstanceOf(Promise);
          const row = await query;
          expect(row).toBeTruthy();
        });
        it('XXXXX should not exist', async () => {
          const query = existeEstacionPorNombre('XXXXX');
          expect(query).toBeInstanceOf(Promise);
          const row = await query;
          expect(row).toBeFalsy();
        });
        it('YYY should exist', async () => {
          const query = existeEstacionPorSigla(SIGLA_ESTACION);
          expect(query).toBeInstanceOf(Promise);
          const row = await query;
          expect(row).toBeTruthy();
        });
        it('XXX should not exist', async () => {
          const query = existeEstacionPorSigla('XXX');
          expect(query).toBeInstanceOf(Promise);
          const row = await query;
          expect(row).toBeFalsy();
        });
      });
      describe('updateEstacion', () => {
        it('change 1 record', async () => {
          const query = updateEstacion(idEstacion, {
            latitud: 99,
          });
          expect(query).toBeInstanceOf(Promise);
          const result = await query;
          expect(result).toBeTruthy();
          const row = await readEstacion(idEstacion);
          expect(row.latitud).toBe(99);
        });
        it('should fail on non-existing record', async () => {
          const query = updateEstacion(0, {
            latitud: 99,
          });
          expect(query).toBeInstanceOf(Promise);
          const result = await query;
          expect(result).toBeFalsy();
        });
      });
      describe('createEstacion', () => {
        it('should fail on duplicate', async () => {
          const query = createEstacion({
            nombre: NOMBRE_ESTACION,
            sigla: SIGLA_ESTACION,
            latitud: 25,
            longitud: 52,
          });
          expect(query).toBeInstanceOf(Promise);
          const result = await query;
          expect(result).toBeFalsy();
        });
      });
    });
  });
});
