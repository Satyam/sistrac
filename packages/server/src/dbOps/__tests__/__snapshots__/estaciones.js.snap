// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbOps.estaciones with mock db createEstacion fail 1`] = `"insert into Estaciones set \`nombre\` = 'Retiro', \`sigla\` = 'RET'"`;

exports[`dbOps.estaciones with mock db createEstacion success 1`] = `"insert into Estaciones set \`nombre\` = 'Retiro', \`sigla\` = 'RET'"`;

exports[`dbOps.estaciones with mock db deleteEstacion fail 1`] = `"delete from Estaciones where idEstacion = 25"`;

exports[`dbOps.estaciones with mock db deleteEstacion success 1`] = `"delete from Estaciones where idEstacion = 25"`;

exports[`dbOps.estaciones with mock db existeEstacionPorNombre fail 1`] = `"select count(*) as hay from Estaciones where nombre = 'Retiro'"`;

exports[`dbOps.estaciones with mock db existeEstacionPorNombre success 1`] = `"select count(*) as hay from Estaciones where nombre = 'Retiro'"`;

exports[`dbOps.estaciones with mock db existeEstacionPorSigla fail 1`] = `"select count(*) as hay from Estaciones where sigla = 'RET'"`;

exports[`dbOps.estaciones with mock db existeEstacionPorSigla success 1`] = `"select count(*) as hay from Estaciones where sigla = 'RET'"`;

exports[`dbOps.estaciones with mock db readEstacion 1`] = `"select * from Estaciones where idEstacion = \`idEstacion\` = 25"`;

exports[`dbOps.estaciones with mock db readEstacionPorNombre 1`] = `"select * from Estaciones where nombre = 'Retiro'"`;

exports[`dbOps.estaciones with mock db readEstacionPorSigla 1`] = `"select * from Estaciones where sigla = 'RET'"`;

exports[`dbOps.estaciones with mock db readEstaciones 1`] = `"select * from Estaciones order by nombre"`;

exports[`dbOps.estaciones with mock db updateEstacion fail  1`] = `"update Estaciones set \`nombre\` = 'Retiro', \`sigla\` = 'RET' where idEstacion = 25"`;

exports[`dbOps.estaciones with mock db updateEstacion success  1`] = `"update Estaciones set \`nombre\` = 'Retiro', \`sigla\` = 'RET' where idEstacion = 25"`;
